backend/reservations/views/admin.py

```diff
@@ -30,127 +30,247 @@
+                  [guest_new["email"]],
+                  auth_user=os.environ['ROOMBAHT_EMAIL_HOST_USER'],
+                  auth_password=os.environ['EMAIL_HOST_PASSWORD'],
+                  fail_silently=False)
+
+def find_room(room_product):
+    room_type = None
+    for a_type, a_products in ROOM_LIST.items():
+        for a_product in a_products:
+            if a_product == room_product:
+                room_type = a_type
+                break
+
+    if not room_type:
+        raise Exception("Unable to actually find room type for %s" % room_product)
+
+    available_room = Room.objects \
+                         .filter(is_available=True,
+                                 name_take3=room_type
+                                 ) \
+                         .order_by('?') \
+                         .first()
+
+    if not available_room:
+        logger.debug("No room of type %s available. Product: %s",
+                     room_type,
+                     room_product)
     else:
-        room.primary=guest.name
+        logger.debug("Found free room of type %s: %s",
+                     room_type,
+                     available_room.number)
 
-    room.save()
+    return available_room
 
-    logger.info("Assigned room type %s (#%s) to %s", room.name_take3, room.number, guest.name)
-    if(email_onboarding):
-        if os.environ.get('ROOMBAHT_SEND_MAIL', 'FALSE').lower() == 'true':
-            time.sleep(5)
-            apppass = os.environ['ROOMBAHT_EMAIL_HOST_PASSWORD']
-            logger.debug(f'[+] Sending invite for guest {guest_new["first_name"]} {guest_new["last_name"]}')
+def reconcile_orphan_rooms(guest_rows):
+    # rooms may be orphaned due to placement changes, data corruption, machine elves
+    def get_guest_obj(name):
+        for guest in guest_rows:
+            if name == f"{guest['first_name']} {guest['last_name']}":
+                return guest
 
-            body_text = f"""
-                BleepBloopBleep, this is the Room Service RoomBaht for Room Swaps letting you know the floors have been cleaned and you have been assigned a room. No bucket or mop needed.
-
-                After you login below you can view your current room, look at other rooms and send trade requests. This functionality is only available until Monday 11/7 at 5pm PST, so please make sure you are good with what you have or trade early.
-
-                Goes without saying, but don't forward this email.
+        return None
 
-                This is your password, there are many like it but this one is yours. Once you use this password on a device, RoomBaht will remember you, but only on that device.
-                Copy and paste this password. Because let’s face it, no one should trust humans to make passwords:
-                {otp}
-                http://rooms.take3presents.com/login
+    orphan_rooms = Room.objects \
+                       .filter(guest=None, is_available=False) \
```

Comment:
@otakup0pe Would we not want to filter for `is_available=False` and also `is_swappable=True`??

Reply:
Placed rooms (manual, art, etc) might not be swappable.